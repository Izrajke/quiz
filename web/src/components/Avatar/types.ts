import type { ReactElement } from 'react';
import type { AvatarFullConfig } from 'react-nice-avatar';

export type FaceColor = '#AC6651' | '#F9C9B6';
export type HairColor = '#000';
export type BgColor = '#E0DDFF';
export type HatColor = '#000';
export type ShirtColor = '#F9C9B6';

export interface AvatarConfig
  extends Required<
    Pick<
      AvatarFullConfig,
      | 'hairStyle'
      | 'hatStyle'
      | 'eyeStyle'
      | 'glassesStyle'
      | 'earSize'
      | 'noseStyle'
      | 'mouthStyle'
      | 'shirtStyle'
      | 'shirtColor'
    >
  > {
  faceColor: FaceColor;
  hairColor: HairColor;
  bgColor: BgColor;
  hatColor: HatColor;
  shirtColor: ShirtColor;
}

export type AvatarConfigWithoutColors = Omit<
  AvatarConfig,
  'faceColor' | 'hairColor' | 'hatColor' | 'bgColor' | 'shirtColor'
>;

type SettingsVariants<T> = Array<{
  value: T;
  view: ReactElement;
}>;

export type SettingsItems = {
  [key in keyof AvatarConfig]: {
    currentValue: AvatarConfig[key];
    allVariants: SettingsVariants<AvatarConfig[key]>;
  };
};
