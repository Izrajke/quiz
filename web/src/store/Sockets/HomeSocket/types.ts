export enum HomeSocketResponseType {
  /** Сообщение из чата */
  chatMessage = 100,
}

export enum HomeSocketRequestType {
  /** Отправить сообщение */
  sendMessage = 10,
}

export interface HomeSocketMessage {
  uuid: string;
  type: HomeSocketResponseType;
  author: string;
  message: string;
  time: number;
}

export interface HomeSocketSendMessage {
  type: HomeSocketRequestType.sendMessage;
  message: string;
}

export type HomeSocketResponse = HomeSocketMessage;

export type HomeSocketRequest = HomeSocketSendMessage;
