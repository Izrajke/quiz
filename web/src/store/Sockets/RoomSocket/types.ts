/** ------------------------------------------------------------ */
/** Типы и интерфейсы сокета */

/** Типы вопросов */
export enum RoomSocketResponseType {
  /** Обычный вопрос с 4-мя вариантами ответа */
  firstQuestionType = 1,
  /** Вопрос без вариантов ответа */
  secondQuestionType = 2,
  /** Ответ на вопрос первого типа */
  answerFirstQuestionType = 5,
  /** Ответ на вопрос второго типа */
  answerSecondQuestionType = 6,
  /** Кол-во территории для захвата */
  allowedToCapture = 7,
  /** Очередь ходов при "захвате" */
  captureTurnQueue = 8,
  /** Очередь ходов при "атаке" */
  attackTurnQueue = 9,
  /** Индекс текущего хода */
  currentTurnIndex = 10,
  /** Информация о игроках */
  playersInfo = 12,
  /** Информация о карте */
  mapInfo = 13,
  /** Начало этапа нападения */
  attackStage = 15,
  /** Конец игры */
  endGame = 999,
}

export enum RoomSocketRequestType {
  /** Отправить ответ */
  sendAnswer = 1,
  /** Получить вопрос */
  getQuestion = 2,
  /** Получить пустую клетку */
  getCell = 3,
  /** Напасть на клетку */
  attackCell = 4,
}

/** - - - - - - - - - - - - - - - - - - - - - - - - */
/** Серверные */

export type PlayerColors = 'player-1' | 'player-2' | 'player-3';

/** Интерфейс объекта вариантов ответ */
export interface RoomSocketOptions {
  [key: string]: string;
}

/** Интерфейс тела вопроса */
export interface RoomSocketQuestion {
  title: string;
  options?: RoomSocketOptions;
}

/** Интерфейс вопроса */
export interface SocketQuestionData {
  type: RoomSocketResponseType;
  question: RoomSocketQuestion;
}

export interface RoomSocketAllowedToCaptureData {
  type: RoomSocketResponseType;
  color: PlayerColors;
  count: number;
}

export interface RoomAnswerOptions {
  name: string;
  color: PlayerColors;
  value: number;
  time: number;
}

/** Порядок ходов стадии "freeCapture/capture" */
export interface RoomTurnQueueData {
  type: RoomSocketResponseType;
  turns: number | PlayerColors[];
}

/** Текущий номер хода */
export interface RoomCurrentTurnData {
  type: RoomSocketResponseType;
  number: number;
}

/** Интерфейс ответа на вопрос */
export interface RoomSocketAnswerData {
  type: RoomSocketResponseType;
  answer: {
    value: string;
  };
  options?: RoomAnswerOptions[];
}

/** Интерфейс игрока */
export interface RoomPlayer {
  id: string;
  name: string;
  points: number;
  color: PlayerColors;
  avatar: string;
}

/** Информация о игроках */
export interface RoomSocketPlayersData {
  type: RoomSocketResponseType;
  players: RoomPlayer[];
}

/** - - - - - - - - - - - - - - - - - - - - - - - - */
/** Карта */

/** Владелец ячейки */
export type RoomCellOwner = PlayerColors | 'empty';

/** Ячейка карты */
export interface RoomCellData {
  /** Доступна ли для игры клетка */
  isExists: boolean;
  /** Владелец */
  owner: RoomCellOwner;
}

/** Row */
export type RoomRowData = RoomCellData[];

/** Карта */
export type RoomMapData = RoomRowData[];

/** Информация о карте */
export interface SocketMapData {
  type: RoomSocketResponseType;
  map: RoomMapData;
}

/** - - - - - - - - - - - - - - - - - - - - - - - - */
/** Клиентские */

/** Интерфейс отправки ответа с клиента */
export interface RoomSocketAnswer {
  type: RoomSocketRequestType;
  option: string;
}

/** Получить вопрос */
export interface RoomSocketGetQuestion {
  type: RoomSocketRequestType;
}

export interface RoomSocketAttackCell {
  type: RoomSocketRequestType;
  rowIndex: number;
  cellIndex: number;
}

/** ------------------------------------------------------------ */

export type RoomSocketResponse = RoomSocketAnswerData &
  RoomSocketPlayersData &
  SocketQuestionData &
  SocketMapData &
  RoomSocketAllowedToCaptureData &
  RoomTurnQueueData &
  RoomCurrentTurnData;

export type RoomSocketRequest =
  | RoomSocketAnswer
  | RoomSocketGetQuestion
  | RoomSocketAttackCell;
