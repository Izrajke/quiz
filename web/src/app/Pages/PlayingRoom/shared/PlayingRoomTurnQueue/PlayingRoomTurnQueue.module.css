.root {
  position: absolute;
  top: 24px;
  left: 24px;
  z-index: 2000;
}
