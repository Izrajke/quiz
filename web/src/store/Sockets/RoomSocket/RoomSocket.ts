import { makeObservable, action } from 'mobx';

import type { RootStore } from 'store/RootStore';
import type { RoomSocketRequest } from 'store/Sockets/RoomSocket/types';

import { RoomSocketResponseController } from './RoomSocketResponseController';

export class RoomSocket {
  root: RootStore;
  private name = 'RoomSocket';
  private url = 'ws://127.0.0.1:8080/ws';
  private socket: WebSocket | undefined;

  controller: RoomSocketResponseController;

  constructor(root: RootStore) {
    makeObservable(this, {
      /** action */
      connect: action,
    });

    this.root = root;
    this.controller = new RoomSocketResponseController(
      this,
      this.root.room,
      this.root.player,
    );
  }

  connect = (roomId: string) => {
    const searchParams = new URLSearchParams({
      room: roomId,
      name: this.root.player.nickname,
    }).toString();

    this.socket = new WebSocket(`${this.url}?${searchParams}`);

    this.socket.onmessage = (evt) => {
      const data = JSON.parse(evt.data);
      this.root.sockets.socketActionRegister('received', data);
      this.controller.control(data);
    };

    this.socket.onclose = () => {
      console.log('Сокет закрылся');
    };

    this.socket.onerror = () => {
      this.root.sockets.connectionError(this.name);
    };
  };

  disconnect = () => {
    if (this.socket) {
      this.socket.close();
    }
  };

  /** Отправить сообщение сокету */
  send(body: RoomSocketRequest) {
    if (this.socket) {
      this.socket.send(JSON.stringify(body));
      this.root.sockets.socketActionRegister('sent', body);
    }
  }
}
