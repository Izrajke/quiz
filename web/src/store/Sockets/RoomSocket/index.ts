export * from './RoomSocket';
export * from './types';
