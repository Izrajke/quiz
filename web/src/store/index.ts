export * from './store';
export * from './Room';
export * from './Player';
export * from './App';
export * from './RootStore';
export * from './Dictionaries';
export * from './Home';
