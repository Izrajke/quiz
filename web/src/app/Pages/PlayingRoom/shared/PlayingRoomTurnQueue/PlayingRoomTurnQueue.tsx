import type { FunctionComponent } from 'react';

import { observer } from 'mobx-react-lite';

import { useStore } from 'store';
import { Attack } from './Attack';
import { Capture } from './Capture';

import classes from './PlayingRoomTurnQueue.module.css';

export const PlayingRoomTurnQueue: FunctionComponent = observer(() => {
  const { room } = useStore();

  if (!room.turnQueue) return null;

  return (
    <div className={classes.root}>
      {room.turnQueue && room.moveStatus !== 'attack' ? (
        <Capture />
      ) : (
        <Attack />
      )}
    </div>
  );
});

PlayingRoomTurnQueue.displayName = 'PlayingRoomTurnQueue';
