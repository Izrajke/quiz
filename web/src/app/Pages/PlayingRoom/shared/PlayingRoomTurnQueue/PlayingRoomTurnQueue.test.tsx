import { Provider } from 'mobx-react';
import { screen, render } from '@testing-library/react';

import { store } from 'store';
import { RoomSocketResponseType } from 'store/Sockets/RoomSocket/types';
import type {
  RoomTurnQueueData,
  RoomCurrentTurnData,
  RoomSocketPlayersData,
} from 'store/Sockets/RoomSocket/types';

import { PlayingRoomTurnQueue } from './PlayingRoomTurnQueue';

describe('INTEGRATION: PlayingRoomTurnQueue', () => {
  const { room } = store;

  const component = (
    <Provider {...store}>
      <PlayingRoomTurnQueue />
    </Provider>
  );

  test('Компонент в контексте capture/freeCapture рендерится корректно', () => {
    render(component);
    const queue: RoomTurnQueueData = {
      type: 8,
      turns: 9,
    };

    const turn: RoomCurrentTurnData = {
      type: 10,
      number: 3,
    };

    room.setCurrentTurn(turn);
    room.setTurnQueue(queue);

    const turnBars = screen.getAllByTestId('TurnBar');
    expect(turnBars.length).toBe(9);

    console.log(room.currentTurn);
    screen.debug();

    expect(turnBars[0].classList.contains('turnBarPast')).toBe(true);
    expect(turnBars[turn.number - 1].classList.contains('turnBarActive')).toBe(
      true,
    );
    expect(turnBars[7].classList.contains('turnBarFuture')).toBe(true);
    expect(turnBars).toMatchSnapshot();
  });

  test('Компонент в конексте attack рендерится корректно', () => {
    render(component);

    const players: RoomSocketPlayersData = {
      type: RoomSocketResponseType.playersInfo,
      players: [
        {
          color: 'player-1',
          name: 'Вася',
          id: '1',
          points: 100,
          avatar: '',
        },
        {
          color: 'player-2',
          name: 'gogsh',
          id: '2',
          points: 200,
          avatar: '',
        },
        {
          color: 'player-3',
          name: 'Петя',
          id: '3',
          points: 300,
          avatar: '',
        },
      ],
    };
    const queue: RoomTurnQueueData = {
      type: RoomSocketResponseType.attackTurnQueue,
      turns: [
        'player-1',
        'player-2',
        'player-3',
        'player-1',
        'player-3',
        'player-2',
      ],
    };
    const turn: RoomCurrentTurnData = {
      type: 10,
      number: 3,
    };

    room.setMoveStatus('attack');
    room.setPlayers(players);
    room.setCurrentTurn(turn);
    room.setTurnQueue(queue);

    const stepContainer = screen.getAllByTestId('StepContainer');
    const turnBars = screen.getAllByTestId('TurnBar');

    expect(stepContainer.length).toBe(2);
    expect(turnBars.length).toBe(6);

    expect(stepContainer).toMatchSnapshot();
    expect(turnBars).toMatchSnapshot();
    expect(turnBars[1].classList.contains('turnBarPast')).toBe(true);
  });
});
