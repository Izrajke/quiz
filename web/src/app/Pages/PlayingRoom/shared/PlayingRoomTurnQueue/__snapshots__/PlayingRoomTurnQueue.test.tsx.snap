// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INTEGRATION: PlayingRoomTurnQueue Компонент в конексте attack рендерится корректно 1`] = `
Array [
  <div
    class="step"
    data-testid="StepContainer"
  >
    <div
      class="turnBar player-1 turnBarPast"
      data-testid="TurnBar"
    />
    <div
      class="turnBar player-2 turnBarPast"
      data-testid="TurnBar"
    />
    <div
      class="turnBar player-3 turnBarActive"
      data-testid="TurnBar"
    />
  </div>,
  <div
    class="step"
    data-testid="StepContainer"
  >
    <div
      class="turnBar player-1"
      data-testid="TurnBar"
    />
    <div
      class="turnBar player-3"
      data-testid="TurnBar"
    />
    <div
      class="turnBar player-2"
      data-testid="TurnBar"
    />
  </div>,
]
`;

exports[`INTEGRATION: PlayingRoomTurnQueue Компонент в конексте attack рендерится корректно 2`] = `
Array [
  <div
    class="turnBar player-1 turnBarPast"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar player-2 turnBarPast"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar player-3 turnBarActive"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar player-1"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar player-3"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar player-2"
    data-testid="TurnBar"
  />,
]
`;

exports[`INTEGRATION: PlayingRoomTurnQueue Компонент в контексте capture/freeCapture рендерится корректно 1`] = `
Array [
  <div
    class="turnBar turnBarPast"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarPast"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarActive"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarFuture"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarFuture"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarFuture"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarFuture"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarFuture"
    data-testid="TurnBar"
  />,
  <div
    class="turnBar turnBarFuture"
    data-testid="TurnBar"
  />,
]
`;
